### PR TITLE
feat(onyx-1071): allow syncing my collection artwork data when creating submission

### PR DESCRIPTION
**Pull request by nickskalkin**
This PR allows to sync artwork data when specifying it's ID as `myCollectionArtworkID` argument to `createConsignmentSubmission` mutation.

Previously, `createConsignmentSubmission` was simply stitched from Convection. This PR adds a resolver for this mutation (keeping the same name). This mutation does two things:

- If  `myCollectionArtworkID` is specified - fetches artwork data using `artworkLoader` and prepares it to be passed to Convection.
- Calls `convectionCreateConsignmentSubmission` mutation (which was renamed from `createConsignmentSubmission`) with artwork data + mutation input arguments (mutation input arguments override artwork data if specified)

Example request (uses this [artwork](https://staging.artsy.net/collector-profile/my-collection/artwork/6654590c43bd0200125ea0b1)):

```graphql
mutation {
  createConsignmentSubmission(input: { artistID: "banksy", myCollectionArtworkID: "6654590c43bd0200125ea0b1" }) {
    consignmentSubmission {
      artistId
      myCollectionArtworkID
      title
      
      medium
      category
      year
      
      attributionClass
      
      editionNumber
      editionSize
      
      height
      width
      depth
      
      dimensionsMetric
      provenance
      
      locationCity
      locationCountry
      locationState
      locationCountryCode
      locationPostalCode
    }
  }
}
```

Response:

```json
{
  "data": {
    "createConsignmentSubmission": {
      "consignmentSubmission": {
        "artistId": "banksy",
        "myCollectionArtworkID": "6654590c43bd0200125ea0b1",
        "title": "Happy Choppers",
        "medium": "Screen print in colours on wove paper",
        "category": "Drawing, Collage or other Work on Paper",
        "year": "2003",
        "attributionClass": "LIMITED_EDITION",
        "editionNumber": "1",
        "editionSize": "2",
        "height": "10",
        "width": "20",
        "depth": "",
        "dimensionsMetric": "in",
        "provenance": "Stolen",
        "locationCity": "Berlin",
        "locationCountry": "Germany",
        "locationState": "Berlin",
        "locationCountryCode": null,
        "locationPostalCode": null
      }
    }
  }
}
```

### CODE REVIEW

File: src/lib/artworkToSubmissionMapping.ts
Comment (dblandin) [inline src/lib/artworkToSubmissionMapping.ts:16]: This makes me cringe a bit knowing that we already have [many versions of this mapping across our repos](https://github.com/search?q=org%3Aartsy+FASHION_DESIGN_AND_WEARABLE_ART&type=code) 😬 

I hope some day soon we won't need a different set of accepted categories between artworks and consignment submissions. Category (Medium Type) is one of those attributes that I'm hoping we can defer fully to the Artwork model and deprecate its presence in the Submission model entirely. 😅

File: src/lib/stitching/convection/v2/stitching.ts
Comment (dblandin) [inline src/lib/stitching/convection/v2/stitching.ts:125]: I think as written, if we currently have clients which execute `createConsignmentSubmission` while passing both `myCollectionArtworkID` and other artwork data input arguments, any input argument with a value of `null` will override any fetched value from the artwork.

Might be fine for this iteration but something to double-check on in Force and Eigen.

```typescript
const artworkSubmissionData = { title: "Penicillin G Procaine" }
const input = { title: null }

const merged = {
  ... artworkSubmissionData,
  ...input,
}
// { title: null }
```

**Review by dblandin — COMMENTED** [review]
Looking good!

File: src/lib/stitching/convection/v2/stitching.ts
Comment (olerichter00) [reply src/lib/stitching/convection/v2/stitching.ts:125]: In Eigen currently, all values would be overridden by the form inputs when submitting a MyC artwork, and in Force, we would create the submission from the MyC artwork without sending any additional (artwork-related) data.

I guess overriding fields with `null` values is okay. It would allow clearing fields in the submission form that have been set on the MyC artwork.

File: src/lib/__tests__/artworkToSubmissionMapping.test.ts
Comment (olerichter00) [inline src/lib/__tests__/artworkToSubmissionMapping.test.ts:39]: This is what I often use in these cases to save time by not having to write down the expected values.

```suggestion
    expect(submission).toMatchInlineSnapshot()
```

File: src/lib/artworkToSubmissionMapping.ts
Comment (olerichter00) [inline src/lib/artworkToSubmissionMapping.ts:22]: 👍

File: src/lib/artworkToSubmissionMapping.ts
Comment (olerichter00) [inline src/lib/artworkToSubmissionMapping.ts:69]: We could use [Lodash's `omit`](https://lodash.com/docs/4.17.15#omit) to filter out undefined values.

```suggestion
  const values = {
    artistID: artwork.artist?.id,
    title: artwork.title,
    year: (artwork.dates || [])[0]?.toString(),
    medium: artwork.medium,
    category: artworkToSubmissionCategory(artwork.category),
    attributionClass: artwork.attribution_class
      ?.replace(" ", "_")
      ?.toUpperCase(),
    editionNumber: artwork.edition_sets?.[0]?.available_editions?.[0],
    editionSize: artwork.edition_sets?.[0]?.edition_size
      ? +artwork.edition_sets?.[0]?.edition_size
      : undefined,
    height: artwork.height,
    width: artwork.width,
    depth: artwork.depth,
    dimensionsMetric: artwork.metric,
    provenance: artwork.provenance,
    locationCity: artwork.collector_location?.city,
    locationCountry: artwork.collector_location?.country,
    locationState: artwork.collector_location?.state,
    locationCountryCode: artwork.collector_location?.countryCode,
    locationPostalCode: artwork.collector_location?.postalCode,
  }

  return omit(values, isUndefined)
```

File: src/lib/stitching/convection/v2/stitching.ts
Comment (olerichter00) [inline src/lib/stitching/convection/v2/stitching.ts:115]: Maybe this is easier to debug 🤷 

```suggestion
                throw new Error(`My Collection artwork (ID: ${myCollectionArtworkID}) not found.`)
```

**Review by olerichter00 — APPROVED** [review]
This looks great 🙌

File: src/lib/artworkToSubmissionMapping.ts
Comment (MounirDhahri) [reply src/lib/artworkToSubmissionMapping.ts:16]: I was under the impression only these are the valid categories we accept
```typescript
{
  OTHER: "OTHER",
  PAINTING: "PAINTING",
  DRAWING_COLLAGE_OR_OTHER_WORK_ON_PAPER: "DRAWING_COLLAGE_OR_OTHER_WORK_ON_PAPER",
  PHOTOGRAPHY: "PHOTOGRAPHY",
  PRINT: "PRINT",
  SCULPTURE: "SCULPTURE",
}
```

File: src/lib/artworkToSubmissionMapping.ts
Comment (MounirDhahri) [inline src/lib/artworkToSubmissionMapping.ts:26]: would this `.replace(/\W/g, "")` work as well?

File: src/lib/artworkToSubmissionMapping.ts
Comment (MounirDhahri) [inline src/lib/artworkToSubmissionMapping.ts:37]: I created a helper here [getInitialSubmissionFormValuesFromArtwork](https://github.com/artsy/eigen/blob/d87a2641244c4c65f853fa47a75796f7915a3975/src/app/Scenes/SellWithArtsy/ArtworkForm/Utils/getInitialSubmissionValuesFromArtwork.ts#L12) for some mapping

File: src/lib/artworkToSubmissionMapping.ts
Comment (MounirDhahri) [inline src/lib/artworkToSubmissionMapping.ts:42]: I am using `artwork.date` in the fronted - not sure if that's the best way

File: src/lib/stitching/convection/v2/stitching.ts
Comment (MounirDhahri) [inline src/lib/stitching/convection/v2/stitching.ts:118]: We need to also specify that the source  is `MY_COLLLECTION` when the artwork comes from my collection.

**Review by MounirDhahri — COMMENTED** [review]
Nice! looks neat. I aded just a few small comments but overall looks good to me.

File: src/lib/artworkToSubmissionMapping.ts
Comment (dblandin) [reply src/lib/artworkToSubmissionMapping.ts:16]: Yeah – and that does align with available artwork mediums defined [here](https://github.com/artsy/metaphysics/blob/9ccad9c469d57897140a20e8bec62af4ff3d3ff9/src/lib/artworkMediums.ts#L4) in Metaphysics.

I think there's some friction between this MP-backed list, which I'd prefer to use in the UI, and the enum defined in Convection. I'm thinking soon we'll be able to defer to the MyC artwork for this medium type and we can simply stop using the [Convection-backed enum](https://github.com/artsy/convection/blob/45f7d5984bc26ad90a9ed8ef92bb28f60facccfd/app/graphql/types/category_type.rb#L4).

File: src/lib/stitching/convection/v2/stitching.ts
Comment (olerichter00) [reply src/lib/stitching/convection/v2/stitching.ts:118]: Good idea! I wonder though whether we should set it in the resolver if myCollectionArtworkId is present or if the client should pass it explicitly as an argument 🤔

Comment (MounirDhahri) [reply src/lib/stitching/convection/v2/stitching.ts:118]: I feel like if we are creating an artwork using a my collection artwork id, we can confidently tell this artwork source is my collection.

File: src/lib/artworkToSubmissionMapping.ts
Comment (nickskalkin) [reply src/lib/artworkToSubmissionMapping.ts:37]: thanks @MounirDhahri. The logic is more or less the same, but the difference is that your code is backed by GraphQL response, and our code uses raw gravity response (artworkLoader). I've checked if our code misses any fields that your code has and I added missing `signature` field.

**Comment by nickskalkin** [comment]
@olerichter00 I've addressed code review requests and this PR can be merged. Let's test on staging carefully after it is merged

File: src/lib/artworkToSubmissionMapping.ts
Comment (MounirDhahri) [reply src/lib/artworkToSubmissionMapping.ts:37]: perfect - thanks

**Review by MounirDhahri — APPROVED** [review]

**Comment by olerichter00** [comment]
> @olerichter00 I've addressed code review requests and this PR can be merged. Let's test on staging carefully after it is merged

@nickskalkin The PR in Force to create a new submission from a MyC artwork is ready for review as well: https://github.com/artsy/force/pull/13998

**Comment by MounirDhahri** [comment]
@olerichter00 @nickskalkin I would also suggest to test things on app since we already send the id there as well

**Comment by nickskalkin** [comment]
I will merge it tomorrow morning, don't want to merge today and get it deployed on prod without me knowing

**Pull request merged**
